[lldb] Realpath symlinks for breakpoints

royitaqi opened the pull request:
# TL;DR

Improve the chance of resolving file/line breakpoints by realpath'ing the support files before doing a second match attempt, with some conditions applied (see "Feature" below).

Telemetries and tests are included.

A working [hello-world example](https://github.com/royitaqi/lldb_demos/blob/main/realpath/README.md).


# Problem / motivation

In some development environment, some source files are symlinked for some reasons, then the symlink files are used in the build. This causes the debug info to contain the path to the symlink files, not the original source files. However, during debugging, the IDE opens and uses the original source file to set breakpoints. The result is a mismatch between the source file's real path v.s. the symlink's path,  causing file/line breakpoints to be unresolved for these files.


# Feature

After the normal file/path matching attempt fails, if the basename of the files match yet the directories don't AND the support file's path matches of the prefixes in the setting (see below), realpath the support file, then reattempt the match.

## Setting

New setting `target.source-realpath-prefixes`: A list of prefixes, which the support file's path has to match in order to be realpath'ed. The default is an empty list, which will cost nothing.

## Telemetries

are added. They can be obtained as a part of the Target's Statistics (e.g. `statistics dump`).
1. The number of realpath attempts made.
2. The number of times such realpath'ing help found a match.

## Tests
are added:

```
ninja lldb-unit-test-deps
tools/lldb/unittests/Utility/UtilityTests --gtest_filter="SupportFileList*"
tools/lldb/unittests/Utility/UtilityTests --gtest_filter="RealpathPrefixes*"

ninja lldb-api-test-deps
bin/llvm-lit -sv ../external/llvm-project/lldb/test/API/functionalities/breakpoint/breakpoint_with_realpath_and_source_map/TestBreakpoint.py
```

royitaqi committed:
[lldb] Allow realpathing support files when resolving file/line breakpoints

Differential Revision: https://phabricator.intern.facebook.com/D60317677

royitaqi committed:
Add unit tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: https://phabricator.intern.facebook.com/D60668386

royitaqi committed:
Add logging and telemetry

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: https://phabricator.intern.facebook.com/D60673160

royitaqi committed:
Address additional comments in diffs

jimingham commented on lldb/include/lldb/Utility/RealpathPrefixes.h:
```diff
@@ -0,0 +1,63 @@
+#include <string>
+#include <vector>
+
+namespace lldb_private {
```
These are all in lldb-forward.h.  We tend not to spot-forward-declare individual classes like this, but just include the forwards file.  You probably should put RealpathPrefixes in there as well, so you don't have to forward declare it in CompileUnit.h

royitaqi replied:
Done.

jimingham commented on lldb/include/lldb/Utility/RealpathPrefixes.h:
```diff
@@ -0,0 +1,63 @@
+  llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> m_fs;
+
+  // The optional Target instance to gather statistics.
+  Target *m_target;
```
This seems dangerous.  How do you know this object will never outlive this target?  Maybe this should be a TargetWP?

royitaqi replied:
Done. Holding a WP, while the getter and setting return/take SP.

jimingham commented on lldb/source/Target/TargetProperties.td:
```diff
@@ -150,6 +150,9 @@ let Definition = "target" in {
     Desc<"The strategy to use when settings breakpoints by file and line. Breakpoint locations can end up being inlined by the compiler, so that a compile unit 'a.c' might contain an inlined function from another source file. Usually this is limited to breakpoint locations from inlined functions from header or other include files, or more accurately non-implementation source files. Sometimes code ...
+  def SourceRealpathPrefixes: Property<"source-realpath-prefixes", "FileSpecList">,
+    DefaultStringValue<"">,
+    Desc<"Realpath any source paths that start with one of these prefixes. If the debug info contains symlinks that don't match the original source file locations that the user will use to set breakpoints, then this setting can help resolve breakpoints correctly.">;
```
You document the wild card characters ("" = match all, "/" = match absolute, but only in the private header file.  You should document them here as well.

jimingham replied:
Also:

contains symlinks that -> contains symlinks which

jimingham replied:
Probably want to mention you only handle directory not file symlinks as well.

royitaqi replied:
Updated. Mentioning:
- File basename has to match.
- Handles both symlinked files and directories.
- The wild cards.

jimingham commented on lldb/source/Utility/FileSpecList.cpp:
```diff
@@ -108,52 +113,85 @@ size_t SupportFileList::FindFileIndex(size_t start_idx,
   const bool full = !file_spec.GetDirectory().IsEmpty();
 
+  // Always start by matching the filename first
+  if (!curr_file.FileEquals(file_spec))
```
You don't handle the case where the symlink also changes the file name?  I guess that's less common.

royitaqi replied:
It's both "less common" and "it can potentially introduce huge performance cost", because now for the same breakpoint file we are checking every support file (that matches the prefix list), comparing to checking only the support file that matches the basename (and also matches the prefix list).

jimingham replied:
That's an okay tradeoff, but you should document that you only support symlinked directories, not symlinked files.

royitaqi replied:
To be exact, it's not only symlinked **directories**, but also symlinked **files** which have the same basename. I will try to make it clear in documentation.

jimingham commented on lldb/source/Utility/FileSpecList.cpp:
```diff
@@ -108,52 +113,85 @@ size_t SupportFileList::FindFileIndex(size_t start_idx,
+    return IsCompatibleResult::kNoMatch;
+
+  // Only compare the full name if the we were asked to and if the current
+  // file entry has the a directory. If it doesn't have a directory then we
```
the a -> a

jimingham commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -192,6 +192,10 @@ TargetStats::ToJSON(Target &target,
   }
   target_metrics_json.try_emplace("sourceMapDeduceCount",
                                   m_source_map_deduce_count);
+  target_metrics_json.try_emplace("sourceRealpathAttemptCount",
```
Should you still count this if realpath_prefixes is nullptr?  In that case you definitely didn't attempt to realpath anything.

royitaqi replied:
The `TargetStatistics` class doesn't really know about `RealpathPrefixes`. It just receives the increments. In the case you mentioned, both counts will simply never be incremented and be reported as zeros.

Are you referring to these two additional fields in the output, in the sense that they waste output size and should be omitted? The output size of `statistics dump` can already be really large esp. in the non-summary mode, so I am not sure if adding two additional counts matter that much, but if you think we should omit then I can make the counters `std::optional<uint32_t>`.

jimingham replied:
I think I was being more simple-minded.  In 99% of the uses of lldb, this function will get a nullptr for the realpathprefixes array, and will be doing no realpath work here.  It would be good if those clients didn't suffer any penalty for this facility they aren't using.  But as it stands they going to pay the cost of sticking a 0 value into some JSON array regardless.  Why do you need to add these two fields?  They can't be required, since if I never set a file and line breakpoint in a session, I'll never enter this code, yet the target statistics still have to work.
I have no problem with adding the fields when they represent work that is being attempted.  But for the most part this is going to be work that's never attempted.

jimingham replied:
I was missing where this work is done.  I don't care about two extra fields set once in the JSON.  This is fine.

jimingham commented on lldb/source/Utility/FileSpecList.cpp:
```diff
@@ -108,52 +113,85 @@ size_t SupportFileList::FindFileIndex(size_t start_idx,
-  const size_t num_files = m_files.size();
-  if (start_idx >= num_files)
-    return UINT32_MAX;
+enum IsCompatibleResult {
```
noMatch and onlyFileMatch are pretty clear, but it would be nice to say exactly what you mean by compatible here.  For instance, I was a little surprised that exactly equal is kCompatible?

jimingham commented on lldb/source/Utility/RealpathPrefixes.cpp:
```diff
@@ -0,0 +1,66 @@
+
+  auto is_prefix = [](llvm::StringRef a, llvm::StringRef b,
+                      bool case_sensitive) -> bool {
+    return case_sensitive ? a.consume_front(b) : a.consume_front_insensitive(b);
```
Why don't you care that this time around that you aren't doing this only with full directory names?

royitaqi replied:
Are you asking why don't we call `is_prefix()` with the directory, and instead we are calling it with the path (directory + basename) while we know that the prefix should only be for the directory?

jimingham replied:
I was asking more about the difference between this `is_prefix` and the `is_suffix` that you defined when first figuring out whether one of the real path prefixes are applied to the incoming path.

royitaqi replied:
Ah got it.

The `is_suffix` and the body of `IsCompatible()` is existing code. I refactored it from originally an inner loop into this function. It seems git wasn't able to figure out which lines are new and which is old.

In general I feel prefixes on partial names generally make sense - ppl sometimes name their directories or files with the same prefix. E.g. we have `lldb/Utility/LLDBAssert.h` and `lldb/Utility/LLDBLog.h`.   But not a big deal, if you think full directory/file names are better/safer, then changing into it is easy.

jimingham replied:
I can't remember offhand what the source path mapping does - that's the most obvious analog.  If that does substitution into partial names when adjusting paths, it seems reasonable this do so as well.  But if it doesn't, then this probably shouldn't either.

royitaqi replied:
That's reasonable.

IIUC, the source path logic:
* Say the source-map is: /foo => /bar
* The user set breakpoint for /barqux/main.cpp:1
* It causes `PathMappingList::ReverseRemapPath()`
* STRING prefix match is successful (/bar is prefix of /barqux/main.cpp)
* qux/main.cpp is the remaining list of components
* Each COMPONENT is added onto the other side of the mapping, the /foo, as separate path components
* The result is /foo/qux/main.cpp

Overall I'd say this logic cares about components of paths. (Side note: It's a bit weird to me that the [prefix match](https://github.com/llvm/llvm-project/blob/main/lldb/source/Target/PathMappingList.cpp#L209) is done as a pure string match.)

Let me change `is_prefix` into component-sensitive, too. Probably by checking if the remaining path is empty or starts with a "/", similar to what the `is_suffix` does.

jimingham commented:
Have you checked to make sure this also works when there is a source-path mapping in place for the files that also differ by realpath vrs. symlink in the part past the part the source-path mapping substitutes?  That might "just work" but it also seems like it might not, so it would be good to have a test for that.

royitaqi commented:
@jimingham 

> Have you checked to make sure this also works when there is a source-path mapping in place for the files that also differ by realpath vrs. symlink in the part past the part the source-path mapping substitutes? That might "just work" but it also seems like it might not, so it would be good to have a test for that.

I checked one scenario and it worked: The user set a breakpoint in file A, which is source-mapped to file B. The line table has file C, which is realpath'ed into file B. Hence a match. See [this example](https://github.com/royitaqi/lldb_demos/blob/main/realpath/README.md#run-with-realpathing-and-source-map).  It works because the source-map is done [here](https://fburl.com/afz6bxlk) on the input file A, while the realpath'ing is done on the support file C, so the two are really orthogonal.

I didn't fully understand your question. Did you have a different scenario in mind?

labath commented on lldb/include/lldb/Utility/RealpathPrefixes.h:
```diff
@@ -0,0 +1,63 @@
+  explicit RealpathPrefixes(const FileSpecList &file_spec_list);
+
+  // Sets an optional filesystem to use for realpath'ing. If not set, the real
+  // filesystem will be used.
+  void SetFileSystem(llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs);
+
+  // Sets an optional Target instance to gather statistics.
+  void SetTarget(Target *target) { m_target = target; }
```
Could these be set directly in the constructor (maybe as optional arguments)? From what I can tell they're only used to override values in tests, and immutable classes are much easier to reason about.

labath commented on lldb/source/Utility/RealpathPrefixes.cpp:
```diff
@@ -0,0 +1,66 @@
+
+#include "lldb/Utility/RealpathPrefixes.h"
+
+#include "lldb/Target/Statistics.h"
+#include "lldb/Target/Target.h"
```
This is a layering violation. Code in Utility should not depend on Target stuff. If you want this to be a Utility class, you should use some kind of dependency inversion technique to break this dependency.

royitaqi replied:
> If you want this to be a Utility class
What is the general guidance of when a class should be in "Target" vs. in "Utility"?

> dependency inversion technique
Like how?

I tried to make an interface `RealpathPrefixStats` which contains the two increment functions, then have `TargetStats` implement this interface, in the hope of the Target can create the `RealpathPrefix` and feed its own `TargetStats` to it as `RealpathPrefixStats` (where `RealpathPrefixStats` only sees `RealpathPrefixStats`, which is also in `Utility`).

However, there is the lifecycle problem Jim pointed out earlier, that what guarantees that the `RealpathPrefixStats` object lives longer than the `RealpathPrefix` object. In order to guarantee this, I will need to make `Target::m_stats` a shared pointer, then give a weak pointer of `RealpathPrefixStats` to `RealpathPrefix`.

Is the above a good plan? Do you see a better approach?


---

FWIW, I did search for the inclusion of Target stuff in Utility, but did it wrong and thought it's a common case. This involves a VS Code bug in "Find in Folder...".

labath replied:
> What is the general guidance of when a class should be in "Target" vs. in "Utility"?

We have a description [here](https://lldb.llvm.org/resources/overview.html), though it leaves a lot of room for interpretation. I think it makes sense for this to be in Utility, but then it should not depend on Target, so that it can be an actual utility that can be reused for more than one purpose.

> 
> > dependency inversion technique
> 
> Like how? Below is my attempt. Haven't updated the patch. Want to get your feedback first by describing it below.
> 
> I tried to make an interface `RealpathPrefixStats` which contains the two increment functions, then have `TargetStats` implement this interface (thus virtual functions), in the hope of the Target can create the `RealpathPrefix` and feed its own `TargetStats` to it as `RealpathPrefixStats` (where `RealpathPrefix` only sees `RealpathPrefixStats`, and both are in `Utility`).
> 
> However, there is the lifecycle problem as Jim pointed out earlier, that what guarantees that the `RealpathPrefixStats` object lives longer than the `RealpathPrefix` object. In order to guarantee this, I will need to make `Target::m_stats` a shared pointer, then give a weak pointer of `RealpathPrefixStats` to `RealpathPrefix`.
> 
> Is the above a good plan? Do you see a better approach?

Maybe (to both questions).

You may not need to make m_stats a shared pointer if you use the little-known-but-very powerful feature of shared_ptr -- the [aliasing constructor](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr). With it, you can create a shared_ptr, which points to a subfield (m_stats), but keeps the whole object alive like so `std::shared_ptr(shared_from_this(), &m_stats)`.

A different approach would be to implement the interfaces you need on the Target object itself.

It may also be possible to keep a record of the statistics on the inside the object itself, and only later add it to the ones in the Target. This would make sense if there's a natural place where you can insert the call to gather the stats (either when the object is destructed, when the stats are requested, etc.).

labath replied:
It may also make sense to move more of the statistics code into Utility as well (provided it does not depend on other things -- which right now I see it does, but it probably can be refactored away). This way you wouldn't need to do the interface thing.

royitaqi replied:
Thanks for all the inspiring options you mentioned, especially the aliasing constructor which I didn't know.

> It may also be possible to keep a record of the statistics on the inside the object itself, and only later add it to the ones in the Target. This would make sense if there's a natural place where you can insert the call to gather the stats (either when the object is destructed, when the stats are requested, etc.).

I went for this approach, because it's the easiest to reason about lifecycle and there *is* a natural location to gather these stats. This can also be easily refactored if/when we move statistics code into Utility.

labath reviewed (commented):
I couple of high-level comments, I don't want to get too involved in the design of this feature.

jimingham commented:
> @jimingham
> 
> > Have you checked to make sure this also works when there is a source-path mapping in place for the files that also differ by realpath vrs. symlink in the part past the part the source-path mapping substitutes? That might "just work" but it also seems like it might not, so it would be good to have a test for that.
> 
> I checked one scenario and it worked: The user set a breakpoint in file A, which is source-mapped to file B. The line table has file C, which is realpath'ed into file B. Hence a match. See [this example](https://github.com/royitaqi/lldb_demos/blob/main/realpath/README.md#run-with-realpathing-and-source-map). It works because the source-map is done [here](https://fburl.com/afz6bxlk) on the input file A, while the realpath'ing is done on the support file C, so the two are really orthogonal.
> 
> I didn't fully understand your question. Did you have a different scenario in mind?
> 
> Also, I have updated the patch according to most of your comments. Only two still open, for which I hope for your clarification.

No, I wanted to make sure that scenario worked and that it had a test (since we want it to keep working).  Looks like you did show that it works, but I didn't see a test for that scenario so that it will keep working.  Did I miss that?

royitaqi committed:
Address most of Jim's comments

royitaqi commented:
> No, I wanted to make sure that scenario worked and that it had a test (since we want it to keep working). Looks like you did show that it works, but I didn't see a test for that scenario so that it will keep working. Did I miss that?

Thanks for confirming the scenario. Yeah I will add a test.

royitaqi committed:
Addressed all comments, EXCEPT the lack of a unit test containing both realpath and source-map

royitaqi committed:
Add unit test for realpath + source-map

royitaqi committed:
Add test case for whole directory name matching and root directory as a special case

royitaqi committed:
Fix format and clean up

royitaqi commented:
Hi @jimingham and @labath,

Thank you for your review so far. They were very helpful comments. I appreciate it.

I *think* I have addressed all of your comments.
When you have the time, it will be great if you can take another look.

Thanks,
Roy

royitaqi commented:
Gentle ping. @jimingham and @labath Is there anything else you want me to improve/fix in this PR?

> I think I have addressed all of your comments.
> When you have the time, I will appreciate you taking another look

cc @clayborg

clayborg reviewed (commented):
I am good with this if Jim and Pavel's issues have been resolved.

bulbazord commented on lldb/include/lldb/Utility/RealpathPrefixes.h:
```diff
@@ -0,0 +1,77 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLDB_CORE_REALPATHPREFIXES_H
```
nit: `LLDB_UTILITY_REALPATHPREFIXES_H`

bulbazord commented on lldb/include/lldb/Utility/RealpathPrefixes.h:
```diff
@@ -0,0 +1,77 @@
+  }
+  uint32_t GetSourceRealpathAttemptCount() const {
+    return m_source_realpath_attempt_count;
+  };
```
nit: stray semicolon, this will introduce a compiler warning.

royitaqi replied:
Thanks for the catch~!
Fixed both.

Interestingly this one with the semicolon didn't give me compiler warning on a fresh build. But yeah it should be removed anyways. Thanks again for your sharp eyes.

bulbazord reviewed (commented):
Interesting. I left a few minor comments, I think Jim and Greg got a lot of the edge cases and bulky feedback out of the way already. :)

royitaqi committed:
Minor fixes

royitaqi commented on lldb/test/API/functionalities/breakpoint/breakpoint_with_realpath_and_source_map/TestBreakpoint.py:
```diff
@@ -0,0 +1,158 @@
+
+
+class BreakpointTestCase(TestBase):
+    NO_DEBUG_INFO_TESTCASE = True
```
Can someone help me double check that this is correct?

royitaqi commented on lldb/test/API/functionalities/breakpoint/breakpoint_with_realpath_and_source_map/TestBreakpoint.py:
```diff
@@ -0,0 +1,158 @@
+        target = self.dbg.CreateTarget(exe)
+        self.assertTrue(target, VALID_TARGET)
+
+    @skipIf(oslist=["windows"])
```
And also this?

royitaqi commented:
Hi @jimingham and @labath ,

Could you do a final pass and see if everything looks good?   I'm hoping to merge this today/tomorrow if possible.

Thanks,
Roy

labath reviewed (commented):
The deps seem fine. I like the approach you've chosen.

I'll leave the rest to other reviewers.

royitaqi commented:
I'm looking at this build break with `PATH_MAX`. Trying to figure out what to use instead.

If anyone knows what to use, please LMK.

Sterling-Augustine commented:
```lrwxrwxrwx 1 saugustine primarygroup 13 Aug 16 15:21 test/API/functionalities/breakpoint/breakpoint_with_realpath_and_source_map/symlink1/foo.h -> ../real/foo.h
-rw-r--r-- 1 saugustine primarygroup 59 Aug 16 15:21 test/API/functionalities/breakpoint/breakpoint_with_realpath_and_source_map/symlink2/foo.h
```
Is it intended for the symlink to have greater permissions than the real file? This is causing some problems on filesystems that are strict about this. I plan to fix shortly, but let me know if that is a mistake.

royitaqi commented:
I assume you meant the "w" permission on group and others (because I suppose the "x" is needed for the symlink to function). It's unintended and by accident. I cannot recall how I created them differently.

Yes, a forward fix from you will be super nice. Thanks for flagging this and offering to fix it!
